File: src/main/scala/com/wikibooks/spark/ch7/scala/MonitoringSample.scala
package com.wikibooks.spark.ch7.scala

import org.apache.spark.sql.streaming.{ OutputMode, StreamingQuery, Trigger }
import org.apache.spark.sql.{ DataFrame, SparkSession }
import org.apache.spark.sql.streaming.StreamingQueryManager

object MonitoringSample {

  def main(args: Array[String]): Unit = {

    val spark = SparkSession
      .builder()
      .appName("KafkaSample")
      .master("local[*]")
      .getOrCreate()

    // StreamingQueryListener 를 이용한 모니터링 
    spark.streams.addListener(new SimpleListener())

    run(spark)
  }

  def run(spark: SparkSession) {

    val df = spark
      .readStream
      .format("kafka")
      .option("kafka.bootstrap.servers", "localhost:9092")
      .option("startingOffsets", "earliest")
      .option("subscribe", "test")
      .load

    val query = df.writeStream
      .format("console")
      .start()

    // StreamingQuery를 이용한 모니터링 
    // monitor(query)

    query.awaitTermination()
    
    query.stop
  }

  def monitor(query: StreamingQuery) {
    new Thread(new Runnable {
      def run() {
        while (true) {
          println("query.lastProgress:" + query.lastProgress)
          query.explain(true)
          Thread.sleep(5 * 1000)
        }
      }
    }).start
  }
}

File: src/main/java/com/wikibooks/spark/ch7/MonitoringSample.java
package com.wikibooks.spark.ch7;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.streaming.StreamingQuery;

public class MonitoringSample {

	public static void main(String[] args) throws Exception {

		SparkSession spark = SparkSession.builder().appName("MonitoringSample").master("local[*]").getOrCreate();

		MonitoringSample obj = new MonitoringSample();

		// StreamingQueryListener 를 이용한 모니터링
		spark.streams().addListener(new SimpleListener());

		obj.run(spark);
	}

	private void run(SparkSession spark) throws Exception {

		Dataset<Row> df = spark.readStream() // streaming 모드일때는 readStream, batch 모드일때는 read
				.format("kafka").option("kafka.bootstrap.servers", "localhost:9092").option("subscribe", "test1,test2")
				.option("startingOffsets", "earliest").load();

		StreamingQuery query = df.selectExpr("cast(value as STRING)").writeStream().format("console").start();

		// StreamingQuery를 이용한 모니터링
		// monitor(query);

		query.awaitTermination();
		query.stop();
	}

	private void monitor(StreamingQuery query) throws Exception {
		new Thread(new Runnable() {
			@Override
			public void run() {
				while (true) {
					System.out.println("query.lastProgress:" + query.lastProgress());
					System.out.print("query.explain(true)");
					query.explain(true);
					try {
						Thread.sleep(5 * 1000);
					} catch (InterruptedException e) {
						e.printStackTrace();
					}
				}
			}
		}).start();
	}
}